"Expected type `[closure@...]`, found fn pointer" when unnecessarily `return`ing a closure

Given [the following code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f194bbd3fb74c407becacd9e6927e0df):

```rust
fn peculiar() -> impl Fn(u8) -> u8 {
    return |x| x + 1
}
```

The current output is:

```
error[E0308]: mismatched types
 --> src/lib.rs:1:36
  |
1 |   fn peculiar() -> impl Fn(u8) -> u8 {
  |  ____________________________________^
2 | |     return |x| x + 1
3 | | }
  | |_^ expected closure, found fn pointer
  |
  = note:    expected type `[closure@src/lib.rs:2:12: 2:21]`
          found fn pointer `fn(u8) -> u8`
```

Ideally the code would compile successfully, but if that is infeasible, it would be nice if the diagnostic contained a hint to remove the `return`. (It's rather head-scratching that the closure is being converted to a fn pointer _even though the return type has also been resolved to be the closure type._)

The `return` is entirely unnecessary here, but I imagine a beginner might include it unthinkingly (as I did) and not be able to even guess that the `return` has anything to do with the type error.

----

The results are the same on stable 1.54.0, beta, and nightly.

If I modify the closure to not be coercible to a function pointer by capturing a variable, then the code compiles without error or relevant warning:

```rust
fn peculiar(y: u8) -> impl Fn(u8) -> u8 {
    return move |x| x + y
}
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"FabianWolff"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_typeck/src/check/coercion.rs
+++ b/compiler/rustc_typeck/src/check/coercion.rs
@@ -941,6 +941,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             exprs.len()
         );
 
+        // The following check fixes #88097, where the compiler erroneously
+        // attempted to coerce a closure type to itself via a function pointer.
         if prev_ty == new_ty {
             return Ok(prev_ty);
         }

--- a/compiler/rustc_typeck/src/check/coercion.rs
+++ b/compiler/rustc_typeck/src/check/coercion.rs
@@ -941,6 +941,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             exprs.len()
         );
 
+        if prev_ty == new_ty {
+            return Ok(prev_ty);
+        }
+
         // Special-case that coercion alone cannot handle:
         // Function items or non-capturing closures of differing IDs or InternalSubsts.
         let (a_sig, b_sig) = {

--- a/src/test/ui/coercion/issue-88097.rs
+++ b/src/test/ui/coercion/issue-88097.rs
@@ -1,0 +1,31 @@
+// In #88097, the compiler attempted to coerce a closure type to itself via
+// a function pointer, which caused an unnecessary error. Check that this
+// behavior has been fixed.
+
+// check-pass
+
+fn peculiar() -> impl Fn(u8) -> u8 {
+    return |x| x + 1
+}
+
+fn peculiar2() -> impl Fn(u8) -> u8 {
+    return |x| x + 1;
+}
+
+fn peculiar3() -> impl Fn(u8) -> u8 {
+    let f = |x| x + 1;
+    return f
+}
+
+fn peculiar4() -> impl Fn(u8) -> u8 {
+    let f = |x| x + 1;
+    f
+}
+
+fn peculiar5() -> impl Fn(u8) -> u8 {
+    let f = |x| x + 1;
+    let g = |x| x + 2;
+    return if true { f } else { g }
+}
+
+fn main() {}

